### PR TITLE
Allow DependentType for Node objects template argument.

### PR DESCRIPTION
**Pull request by python3kgae**
Allow DependentType for case like
 template < typename T > void foo(GroupNodeInputRecords< T > data) {}

Also make sure specialization is done before check template argument is not empty struct for case like
template < typename T > void bar(GroupNodeInputRecords< Record < T > > data) {}

Fixes https://github.com/microsoft/DirectXShaderCompiler/issues/5822

### CODE REVIEW

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 1]
```diff
@@ -5377,6 +5377,10 @@ class HLSLExternalSource : public ExternalSemaSource {
         return true;
       }
       QualType argType = arg.getAsType();
+      // Allow TemplateTypeParmType for case like
+      // template <typename T> void foo(GroupNodeInputRecords<T> data) {}
+      if (argType->getAs<TemplateTypeParmType>())
```
Comment (tex3d) [inline tools/clang/lib/Sema/SemaHLSL.cpp:5382]: Couldn't we check for `argType->isDependentType()` instead?

Comment (python3kgae) [reply tools/clang/lib/Sema/SemaHLSL.cpp:5382]: Yes. isDependentType works.
Hit other issue in test.
Debugging.

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 2]
```diff
@@ -5377,6 +5377,26 @@ class HLSLExternalSource : public ExternalSemaSource {
         return true;
       }
       QualType argType = arg.getAsType();
+      // Allow DependentType for case like
+      // template <typename T> void foo(GroupNodeInputRecords<T> data) {}
+      if (argType->isDependentType())
+        return false;
```
Comment (llvm-beanz) [inline tools/clang/lib/Sema/SemaHLSL.cpp:5383]: This will need to be validated during instantiating right? Can you add tests for failure cases?

Comment (python3kgae) [reply tools/clang/lib/Sema/SemaHLSL.cpp:5383]: Done.

File: tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl
Comment (tex3d) [inline tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl:1]: Couldn't these be added to the existing `template-diag-deferred.hlsl`?

File: tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// GroupNodeInputRecords may not non-struct/class DependentType. 
+
+template<typename T>
+struct Record {  }; // expected-note{{zero sized record defined here}}
+
+template <typename T> void foo(GroupNodeInputRecords<T> data) {}
+
+template <typename T> void bar() {
+  GroupNodeInputRecords<Record<T> > data; // expected-error{{record used in GroupNodeInputRecords may not have zero size}}
+  foo(data);
```
Comment (tex3d) [inline tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl:13]: None of the calls to foo() are causing the diagnostic to fire.  I guess it's because it already fired for the local, invalidating the `data` decl?  What if something constructs a local method with an invalid argument type based on the template type, without having called that method.  For example:

```hlsl
template <typename T>
class CFoo {
  T var;
  void foo(GroupNodeInputRecords<T> data) {}
};
// ...
  CFoo<int> c;
```

Should that result in a diagnostic triggered by the `CFoo<int> c` instantiation?

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 4]
```diff
@@ -5380,20 +5380,32 @@ class HLSLExternalSource : public ExternalSemaSource {
       }
 
       QualType ArgTy = Arg.getAsType();
+      // Allow DependentType for case like
+      // template <typename T> void foo(GroupNodeInputRecords<T> data) {}
+      if (ArgTy->isDependentType())
```
Comment (tex3d) [inline tools/clang/lib/Sema/SemaHLSL.cpp:5385]: Is the `case AR_TOBJ_DEPENDENT` in `DiagnoseNodeStructArgument` still hit with this code?  If so, in what case?  Is there a test here that hits it?

**Review by tex3d — APPROVED** [review]
Looks good, some additional thoughts, nothing blocking.

File: tools/clang/lib/Sema/SemaHLSL.cpp (hunk 4 above)
Comment (python3kgae) [reply tools/clang/lib/Sema/SemaHLSL.cpp:5385]: Added.

File: tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl (hunk 3 above)
Comment (python3kgae) [reply tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl:13]: Added.

File: tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl
Comment (python3kgae) [reply tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl:1]: Cannot find VerifierTest which uses template-diag-deferred.hlsl. :(
And lit Shell test is easier to add.

File: tools/clang/lib/Sema/SemaHLSL.cpp (hunk 1 above)
Comment (bob80905) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: Was the issue resolved?

Comment (python3kgae) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: Yes. It is resolved.

File: tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl
Comment (pow2clk) [reply tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl:1]: Ugh. We really need to audit all those unused test files and get them hooked up

Comment (python3kgae) [reply tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl:1]: template-diag-deferred.hlsl was added in https://github.com/microsoft/DirectXShaderCompiler/commit/baae5dd1e52f5da15e829a3199e17499cb9e55c6, but never added for taef test.
Maybe we should just add it as lit Shell test in a separate PR?

File: tools/clang/test/HLSLFileCheck/hlsl/workgraph/template.hlsl
Comment (llvm-beanz) [inline tools/clang/test/HLSLFileCheck/hlsl/workgraph/template.hlsl:1]: Can we make these test file names more descriptive?

File: tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl
Comment (llvm-beanz) [reply tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl:1]: > Maybe we should just add it as lit Shell test in a separate PR?

If this is going to be a separate PR, can you file an issue to track it so that we don't forget about it?

File: tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+template <typename T>
+class CBar {
+  T var;
+  void bar(GroupNodeInputRecords< Texture2D<T> > data) {} // expected-error{{'Texture2D<float>' is not valid as a node record type - struct/class required}}
+};
+
+void testCBar() {
+  CBar<float> c; // expected-note{{in instantiation of template class 'CBar<float>' requested here}}
+}
```
Comment (llvm-beanz) [inline tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl:45]: Can we also add a test where being zero-sized is based on the template to ensure that instantiation is happening?

Something like:
```c++
struct ZeroSize {};

template<typename T>
struct RecordTemplate { T Value; }; // expected-note{{zero sized record defined here}}

template <typename T> void oopsie() {
  GroupNodeInputRecords<RecordTemplate<ZeroSize> > data; // expected-error{{record used in GroupNodeInputRecords may not have zero size}}
  foo(data);
}
```

Additionally we should verify no diagnostic is emitted for something like:

```c++
template <typename T> void woo() {
  GroupNodeInputRecords<RecordTemplate<int> > data;
  foo(data);
}
```

**Review by llvm-beanz — COMMENTED** [review]
A few minor comments, but mostly looks good.

File: tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl
Comment (python3kgae) [reply tools/clang/test/SemaHLSL/hlsl/workgraph/template.hlsl:1]: Added https://github.com/microsoft/DirectXShaderCompiler/issues/5870 to track it.

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 6]
```diff
@@ -5380,20 +5380,32 @@ class HLSLExternalSource : public ExternalSemaSource {
       }
 
       QualType ArgTy = Arg.getAsType();
+      // Allow DependentType for case like
+      // template <typename T> void foo(GroupNodeInputRecords<T> data) {}
```
[the remark this answers is not in the document]
Comment (llvm-beanz) [reply tools/clang/lib/Sema/SemaHLSL.cpp:5384]: Instead of applying my suggestion, you kept the old comment and the new comment, which seems confusing.

File: tools/clang/test/HLSLFileCheck/hlsl/workgraph/template.hlsl
Comment (llvm-beanz) [inline tools/clang/test/HLSLFileCheck/hlsl/workgraph/template.hlsl:1]: You renamed one test file, but they all had bad names. Can you rename the other ones too?

File: tools/clang/lib/Sema/SemaHLSL.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (python3kgae) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: :( Updated. It looks the same in github email.

**Review by llvm-beanz — APPROVED** [review]
